timers: do less work in insert

apapirovski opened the pull request:
Most of the code in insert is only applicable to scheduling non-timers or re-scheduling timers. We can skip most of it in the case of setTimeout, setInterval & setUnrefTimeout.

Second commit fixes a long-standing bug with `refresh` in cases where it's triggered after the timer already ran its course. Prior to this the refreshed timer would end up always being unrefed, even when it was previously refed.

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

Fishrock123 commented on lib/internal/timers.js:
```diff
@@ -289,31 +291,47 @@ function decRefCount() {
 // list if one does not already exist for the specified timeout duration.
-function insert(item, refed, start) {
-  let msecs = item._idleTimeout;
+function schedule(item, refed, start) {
```
IMO we should call this `insertGuarded`.

... and then possibly rename `insert` to `schedule` for both. (May cause too many backport conflicts?)

Fishrock123 commented on lib/internal/timers.js:
```diff
@@ -534,16 +539,14 @@ function getTimerCallbacks(runNextTicks) {
-          if (start === undefined)
-            start = getLibuvNow();
-          insert(timer, timer[kRefed], start);
+          schedule(timer, timer[kRefed], start);
```
This can't be `insert`? (Maybe deserves a comment?)

apapirovski commented:
@Fishrock123 I've updated with some significant changes in the second commit. There was a bug with `refresh` that is now fixed and there's a test case provided.

BridgeAR commented:
Benchmark with the latest code: https://ci.nodejs.org/view/Node.js%20benchmark/job/benchmark-node-micro-benchmarks/356/

Results with any significant change:

```

                                                             confidence improvement accuracy (*)    (**)   (***)
timers/immediate.js type='clear' n=5000000                          ***     19.14 %       ±2.12%  ±2.80%  ±3.60%
timers/immediate.js type='depth' n=5000000                            *      0.83 %       ±0.73%  ±0.96%  ±1.23%
timers/set-immediate-depth-args.js n=5000000                          *     -0.92 %       ±0.78%  ±1.03%  ±1.32%
timers/timers-cancel-unpooled.js direction='end' n=1000000            *     15.14 %      ±12.87% ±16.98% ±21.81%
timers/timers-depth.js n=1000                                         *      0.06 %       ±0.05%  ±0.07%  ±0.09%
timers/timers-insert-unpooled.js direction='end' n=1000000          ***     24.38 %       ±3.03%  ±4.01%  ±5.15%
timers/timers-timeout-nexttick.js n=50000                            **     -4.00 %       ±2.41%  ±3.18%  ±4.08%
timers/timers-timeout-unpooled.js n=1000000                         ***     -4.59 %       ±2.62%  ±3.45%  ±4.44%
```

Fishrock123 commented on lib/timers.js:
```diff
@@ -60,13 +61,14 @@ const {
 
 // Remove a timer. Cancels the timeout and resets the relevant timer properties.
 function unenroll(item) {
+  if (item._destroyed)
```
I'm not sure this is 'correct'...

At very least, `unenroll()` should _always_ call `L.remove(item)`.

That being said, the definitive check for if something is an active timer has been (roughly) kept to `item._idleTimeout > 0 && typeof item._idleCallback === 'function'` for some time (partially since inception), and I think we should avoid changing that assumption if possible?

I feel like the old behavior was more foolproof.

apapirovski replied:
I mean, the goal is to establish a lifecycle where `_destroyed` or something else can sufficiently indicate a timer/object that doesn't need to have anything done to it. It's possible we can (and should) just do it with the check you mentioned.

In addition, the notion of there being an undefined `kRefed` and `null` one is very complex and also means the state of the timer is lossy. I think it's important to get away from that current implementation and towards something where kRefed doesn't need to be unset when a timer expires.

Fishrock123 commented:
@apapirovski thoughts?

apapirovski commented:
@Fishrock123 Your comment is fair. Let's discuss in person at the summit. Didn't have time for the past month to think about this but will probably try to wrap up this PR in the next two days.

Fishrock123 commented:
@apapirovski good to merge?

apapirovski commented:
@fishrock123 yeah, should be. I’ll confirm this weekend and do it if so.

apapirovski commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/27640/

BridgeAR commented:
@apapirovski the latest benchmark results do not indicate a big difference at all and these are the only three benchmarks that seemed to have any significance:

```
01:16:52  timers/timers-depth.js n=1000                                       ***      0.71 %       ±0.19%  ±0.25%  ±0.32%
01:16:52  timers/timers-insert-unpooled.js direction='end' n=1000000            *     -2.55 %       ±2.40%  ±3.17%  ±4.07%
01:16:52  timers/timers-timeout-pooled.js n=10000000                           **     -5.74 %       ±4.14%  ±5.46%  ±7.02%
```

Fishrock123 commented:
I think this change aims more to reduce potential edge-cases than desiring a perf impact?

Fishrock123 commented:
Aside form an extra function closure the only thing I can think of is the coalescing boolean checks...

Trott commented:
Should this be squashed into one commit or two? If two, which commit should the fixup commit go with?

Trott commented:
> Should this be squashed into one commit or two? If two, which commit should the fixup commit go with?

@apapirovski ^^^^^^^^

BridgeAR commented:
@Trott the fixup is for the second commit. I went ahead and landed it in two commits.

Landed in 0b89761d6dcaa1aa6b49b5b9d85b36065f681351, 1fab8a92974ce555adca39baada1d199b4952fd7 :tada:
